compiler/rustc_hir_analysis/src/check/wfcheck.rs

```diff
@@ -1925,6 +1929,139 @@ fn error_392(
     err
 }
 
+fn check_for_redundant_lifetimes<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    owner_id: LocalDefId,
+    outlives_env: &OutlivesEnvironment<'tcx>,
+) {
+    let def_kind = tcx.def_kind(owner_id);
+    match def_kind {
+        DefKind::Struct
+        | DefKind::Union
+        | DefKind::Enum
+        | DefKind::Trait
+        | DefKind::TraitAlias
+        | DefKind::Fn
+        | DefKind::Const
+        | DefKind::Impl { of_trait: false } => {
+            // Proceed
+        }
+        DefKind::AssocFn | DefKind::AssocTy | DefKind::AssocConst => {
+            let parent_def_id = tcx.local_parent(owner_id);
+            if matches!(tcx.def_kind(parent_def_id), DefKind::Impl { of_trait: true }) {
+                // Don't check for redundant lifetimes for trait implementations,
+                // since the signature is required to be compatible with the trait.
+                return;
+            }
+        }
+        DefKind::Impl { of_trait: true }
```

Comment:
why should we not check trait implementations? they can freely dedup lifetimes, can't they?

Reply:
They can dedup only late bound lifetimes. early bound generics must match one-to-one.

Reply:
Specifically they cannot dedup early lifetimes since they must be able to freely rebase substs.

We could only suggest late lifetimes as victim lifetimes for this lint, but I don't think it's worth the extra complication.

Reply:
> I am talking about lifetimes of the impl itself, not of the impl item, e.g.

This is still a bit concerning but for a different reason as changing the impl to silence the lint can be a breaking change. The impls `impl<'a> Trait for &'static &'a u8` and `impl Trait for &'static &'static u8` are identical but only the first satisfies the bound `for<'a> &'static &'a u8: Trait`.

Reply:
> changing the impl to silence the lint can be a breaking change

This is an issue for functions as well as they can break HRTB bounds of the `Fn` trait. 

Not sure how significant this is but at least it tells us to proceed cautiously when bumping the default level of the lint.

Reply:
we also have the same issue with free functions as well. `fn foo<'a>(x: &'static &'a ()) {}` implements `for<'a> fn(&'static &'a ())`, but `fn foo(x: &'static &'static ())` does not.
```rust
fn foo<'a>(x: &'static &'a ()) {} // ok
// fn foo(x: &'static &'static ()) {} // error

fn main() {
    let x: for<'a> fn(&'static &'a ()) = foo;
}
```
I think we should ignore that issue. I don't expect people to encounter that issue in practice